### PR TITLE
Add ovirt as a platform

### PR DESCRIPTION
**Pull request by rgolangh**
Adding the files needed to make ovirt a provider of configs.

Fixes: #766

Added manifests and templates for ovirt platform. This is the same solution like `baremetal` uses for networking so the yamls are almost identical.

 **- What I did**

To test this I'm running `make update && make image` and pushing the container to quay.io. Then I modify my IPI installer to pull this image during bootstrap.

Related: openshift/enhancements#61

**- Description for the changelog**
Adding the manifest, templates and controller handling for oVirt platform.

### CODE REVIEW

**Comment by cgwalters** [comment]
https://github.com/openshift/api/pull/314
needs to land first so:

/hold

Also it looks like you have some unnecessary `.gitkeep` stuff that's no longer needed after: https://github.com/openshift/machine-config-operator/pull/653

File: pkg/controller/template/test_data/controller_config_ovirt.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  clusterDNSIP: "10.3.0.10"
+  cloudProviderConfig: |
+    testing
+    multi-line cloud config
+    [test]
+      option = dummy
+  etcdDiscoveryDomain: "my-test-cluster.metalkube.org"
+  etcdInitialCount: 3
+  platform: "ovirt"
+  etcdCAData: OHVtbXkgZXRjZC1jYQo=
```
Comment (ashcrow) [inline pkg/controller/template/test_data/controller_config_ovirt.yaml]: Just curious, why a different dummy encoded string versus the others?

File: pkg/controller/template/test_data/controller_config_ovirt.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+apiVersion: "machineconfigurations.openshift.io/v1"
+kind: "ControllerConfig"
+spec:
+  clusterDNSIP: "10.3.0.10"
+  cloudProviderConfig: |
+    testing
+    multi-line cloud config
+    [test]
+      option = dummy
+  etcdDiscoveryDomain: "my-test-cluster.metalkube.org"
```
Comment (ashcrow) [inline pkg/controller/template/test_data/controller_config_ovirt.yaml]: Looks like `metalkube.org` is an owned domain. Just checking this is meant to be used as a default/dummy instead of `my-test-cluster.installer.team.coreos.systems`.

File: hack/build-image.sh [hunk 3]
```diff
@@ -64,5 +65,5 @@ for IMAGE_TO_BUILD in $TOBUILD; do
 	NAME="${IMAGE_TO_BUILD#Dockerfile.}"
   NAME="${NAME//.upstream}"
 	set -x
-	$podman build -t "localhost/${NAME}:latest" -f "${IMAGE_TO_BUILD}" --no-cache
+	$podman build -t "localhost/${NAME}:latest" -f "${IMAGE_TO_BUILD}"
```
Comment (runcom) [inline hack/build-image.sh]: Uhm, what about the changes in this file

File: pkg/apis/machineconfiguration.openshift.io/v1/types.go [hunk 4]
```diff
@@ -94,6 +94,10 @@ type ControllerConfigSpec struct {
 
 	// proxy holds the current proxy configuration for the nodes
 	Proxy *configv1.ProxyStatus `json:"proxy"`
+
+	// infra holds the infrastructure details
+	// TODO this makes platform redundant as everything is contained inside Infra.Status
+	Infra configv1.Infrastructure
```
Comment (runcom) [inline pkg/apis/machineconfiguration.openshift.io/v1/types.go]: Use a pointer

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go [hunk 5]
```diff
@@ -132,6 +135,8 @@ func New(
 		DeleteFunc: ctrl.imageConfDeleted,
 	})
 
+	ctrl.infraLister = infraInformer.Lister()
```
Comment (runcom) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go]: Why the lister here, it's not used

File: cmd/machine-config-controller/start.go [hunk 6]
```diff
@@ -121,6 +121,7 @@ func createControllers(ctx *controllercommon.ControllerContext) []controllercomm
 			ctx.ClientBuilder.KubeClientOrDie("container-runtime-config-controller"),
 			ctx.ClientBuilder.MachineConfigClientOrDie("container-runtime-config-controller"),
 			ctx.ClientBuilder.ConfigClientOrDie("container-runtime-config-controller"),
+			ctx.ConfigInformerFactory.Config().V1().Infrastructures(),
```
Comment (runcom) [inline cmd/machine-config-controller/start.go]: Not needed?

File: pkg/operator/bootstrap.go [hunk 7]
```diff
@@ -117,8 +117,6 @@ func RenderBootstrap(
 		templatectrl.KubeClientAgentImageKey: imgs.KubeClientAgent,
 	}
 
-	config := getRenderConfig("", string(filesData[kubeAPIServerServingCA]), spec, &imgs.RenderConfigImages, infra.Status.APIServerInternalURL)
```
Comment (runcom) [inline pkg/operator/bootstrap.go]: Uhm?

File: pkg/operator/cluster-infrastructure.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+status:
+  apiServerInternalURI: https://api-int.blue.metalkube.org:6443
+  apiServerURL: https://api.blue.metalkube.org:6443
+  etcdDiscoveryDomain: blue.metalkube.org
+  infrastructureName: blue-z55jb
+  platform: oVirt
+  platformStatus:
+    ovirt:
+      api_vip: 10.35.0.19
+      dns_vip: 10.35.0.19
```
Comment (runcom) [inline pkg/operator/cluster-infrastructure.yaml]: Why this and the install Config above?

File: pkg/operator/render.go [hunk 9]
```diff
@@ -84,13 +84,16 @@ func createDiscoveredControllerConfigSpec(infra *configv1.Infrastructure, networ
...
+	case configv1.OvirtPlatformType:
+		platform = "ovirt"
 	}
 
 	ccSpec := &mcfgv1.ControllerConfigSpec{
 		ClusterDNSIP:        dnsIP,
 		CloudProviderConfig: "",
 		EtcdDiscoveryDomain: infra.Status.EtcdDiscoveryDomain,
 		Platform:            platform,
+		Infra:				 infra,
```
Comment (runcom) [inline pkg/operator/render.go]: Need to gofmt the file

File: pkg/operator/cluster-infrastructure.yaml (hunk 8 above)
Comment (rgolangh) [reply pkg/operator/cluster-infrastructure.yaml]: junk from past attempts

File: hack/build-image.sh (hunk 3 above)
Comment (rgolangh) [reply hack/build-image.sh]: Not directly related to the patch. But this lets a developer build more quickly the image. Maybe I'll send it as a separate productivity patch?

**Comment by runcom** [comment]
Just for reference, this is an example of how a platform has been added after support has landed into the installer https://github.com/openshift/machine-config-operator/pull/907 (your PR here also adds templates but that's ok ofc)

File: pkg/operator/cluster-config.yaml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        hostPrefix: 23
+      machineCIDR: 10.0.0.0/16
+      networkType: OpenShiftSDN
+      serviceNetwork:
+      - 172.30.0.0/16
+    platform:
+      ovirt:
+        api_floating_ip: 10.35.0.19
+        ovirt_password: "123"
+        ovirt_url: https://rgolan.usersys.redhat.com:8443/ovirt-engine/api
```
Comment (kikisdeliveryservice) [inline pkg/operator/cluster-config.yaml]: is this supposed to be here?

File: pkg/operator/cluster-config.yaml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        ovirt_username: admin@internal
+    pullSecret: ""
+    sshKey: |
+      ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDX95Yb4UAiBLh+6KAchctC7VrUGghC7OLyJJctkA+wxOL3gBSUz1Y95J+PSh1nkuftKGM2Nl4vPULq9UgfpygjetLuiDI4x59IGCmaAKmvTr+moT/DsteIIj2P4JVTxEFFwF55tp4n8Yypd4up20yhj737FmO1AOzjaHHDodGQiXn6cJy5Kd+vn0HSTi6nkR2pyodBWRhV58swAAc3OkjAG2SqCLxuvT1UdXgAY1TJfnE6WBzOGcJ9KyvSFK0M9ROfnsXEvDXDo3zzz0hSWDqH6NqjBVMSULxB143ek80A2ShtcQjV5Xpdu7RBzoA/cFvRBz2pHt3XgwyrJybRNepAS8HzHQ2D/rcsRZ45QnSeHWx4xOhFKWFJ9Ohr9mdX8WkxGVhlavsEZzJndjxGKEMa4HIZ5jm8Y8VJ3PG6kcDfquE7q4h0EopNc2PSRVKX4WhPFT1CmerRerCJAlTegS2VMkCeBiLlfm7if3PWNWKUfH4+4eOGayHoazQdMWwqi4ryV1SwxhrBCFimBTARbufwMIwWf6kDEETpu93+Lpb7/m0JdAKghlSbUqS3YdV0IW73fGRfvj9NpyiUlGkgT4Z1HtHiNWKWhPxeUAtKncPjCI7YUYb1UPLvRh7IrhQVbvdajFzSTFC/e0bstbYIE2KYYjeuPme85wjv+Jqd76O7Ww== rgolan@redhat.com
```
Comment (kikisdeliveryservice) [inline pkg/operator/cluster-config.yaml]: same as above

Comment (rgolangh) [reply pkg/operator/cluster-config.yaml]: thanks for catching this - removed

**Comment by rgolangh** [comment]
I'll add another patch to reduce the cyclo complexity of pkg/operator/bootstrap.go

File: pkg/operator/render.go [hunk 12]
```diff
@@ -14,7 +14,6 @@ import (
 	"github.com/golang/glog"
 
 	configv1 "github.com/openshift/api/config/v1"
-
 	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
```
Comment (kikisdeliveryservice) [inline pkg/operator/render.go]: can you revert this change as well since it's an autospacing?

File: pkg/operator/cluster-config.yaml (hunk 11 above)
Comment (kikisdeliveryservice) [reply pkg/operator/cluster-config.yaml]: no prob, figured you didnt want to share it with the world :smile:

File: pkg/operator/render.go (hunk 12 above)
Comment (rgolangh) [reply pkg/operator/render.go]: done

**Comment by kikisdeliveryservice** [comment]
Looks good now, just want to see e2e tests turn green.

**Comment by kikisdeliveryservice** [comment]
@rgolangh we'll finish reviewing early next week. 

@runcom PTAL

/assign @runcom

**Comment by runcom** [comment]
needs a rebase now :(

File: manifests/ovirt/keepalived.yaml [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    - "-f"
+    - "/etc/keepalived/keepalived.conf"
+    - "--dont-fork"
+    - "--vrrp"
+    - "--log-detail"
+    - "--log-console"
+    resources:
+      requests:
+        cpu: 150m
+        memory: 1Gi
```
Comment (tux-o-matic) [inline manifests/ovirt/keepalived.yaml]: Is that 1GB of RAM just for a Keepalive instance?

File: templates/common/ovirt/files/ovirt-coredns.yaml [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        securityContext:
+          privileged: true
+        image: {{.Images.corednsImage}}
+        args:
+        - "--conf"
+        - "/etc/coredns/Corefile"
+        resources:
+          requests:
+            cpu: 150m
+            memory: 1Gi
```
Comment (tux-o-matic) [inline templates/common/ovirt/files/ovirt-coredns.yaml]: Sounds a little too much to request 1Gi of RAM for a CoreDNS instance.

Comment (rgolangh) [reply templates/common/ovirt/files/ovirt-coredns.yaml]: Thanks for the question, worth finding out. BTW those requested resources are a hint for the scheduler as far as I understand this. And this is a static pod, which means this 
is probably number is meaningless and I should better just drop it @celebdor @cybertron  thoughts?

Comment (tux-o-matic) [reply templates/common/ovirt/files/ovirt-coredns.yaml]: For things meant to run on "infra" VM, the scheduler will have little to choose from so not that relevant to request resources. If you expect a service could leak memory, you could set a memory limit to force a restart in such an invent.
Would reduce the number of lines of YAML to get rid of all those requests on infra Pods.

**Comment by rgolangh** [comment]
> Will defer to @tux-o-matic @celebdor @cybertron to final approve as there is an open issue here: [#767 (comment)](https://github.com/openshift/machine-config-operator/pull/767#discussion_r334785278)
> 
> And it should be resolved/settled.

Uploading a version which drops the resource definition - its irrelevant for static pods and have nothing to do with it being on infra VMs or not.

**Comment by kikisdeliveryservice** [comment]
btw: I dont know where ci stands, so might have to be patient on it merging.

**Comment by kikisdeliveryservice** [comment]
ah didnt realize i did the initial approval. :(

@runcom PTAL

/lgtm cancel

**Comment by runcom** [comment]
looks sane to me

/lgtm
